sky/serve/core.py

```diff
@@ -646,6 +656,180 @@ def status(
+        logger.info('Downloading the replica logs from %s to %s...',
+                    remote_service_dir_name, sky_logs_directory)
+        runner.rsync(source=dir_for_download,
+                     target=sky_logs_directory,
+                     up=False,
+                     stream_logs=False)
+
+        # Remove the logs from the remote server after downloading.
+        logger.info('Preparing to remove replica logs from remote server...')
+        remove_code = (
+            serve_utils.ServeCodeGen.remove_replica_logs_for_download(
+                service_name, run_timestamp))
+        logger.debug('Generated remove code: %s', remove_code)
+        remove_returncode = backend.run_on_head(controller_handle,
+                                                remove_code,
+                                                require_outputs=False,
+                                                stream_logs=False)
+        logger.debug('Remove command return code: %s', remove_returncode)
+
+        try:
+            subprocess_utils.handle_returncode(
+                remove_returncode, remove_code,
+                'Failed to remove the replica logs for '
+                'download on the controller.')
+            logger.info('Replica logs removed successfully'
+                        ' from remote server.')
+        except exceptions.CommandError as e:
+            logger.error('Error occurred while '
+                         'removing replica logs: %s', e.error_msg)
+            raise RuntimeError(e.error_msg) from e
+
+    # We download the replica logs first because that download creates
+    # the timestamp directory, and we can just put the controller and
+    # load balancer logs in there. Otherwise, we would create the
+    # timestamp directory with controller and load balancer logs first,
+    # and then the replica logs download would overwrite it.
+    target_directory = os.path.join(sky_logs_directory, run_timestamp)
+    os.makedirs(target_directory, exist_ok=True)
+    single_file_synced = False
+
+    if (sync_down_all_components or
+            service_component == serve_utils.ServiceComponent.CONTROLLER):
+        logger.info(
+            'Starting the process to prepare and download controller logs...')
+        runner = controller_handle.get_command_runners()
+        controller_log_file_name = (
+            serve_utils.generate_remote_controller_log_file_name(service_name))
+        logger.info('Downloading the controller logs...')
+        runner.rsync(source=controller_log_file_name,
+                     target=os.path.join(
+                         target_directory,
+                         serve_constants.CONTROLLER_LOG_FILE_NAME),
+                     up=False,
+                     stream_logs=False)
+        if not sync_down_all_components:
+            single_file_synced = True
+            target_directory = os.path.join(
+                target_directory, serve_constants.CONTROLLER_LOG_FILE_NAME)
```

Comment:
is it possible that we keep a list of source & target and use loop to rsync them down?

Reply:
Got it, just to confirm:

In fact, `sync_down_logs()` for the controller/load balancer only deals with one instance. Unlike replicas, which have multiple instances where using a loop is more convenient.

So, the reason for "keeping a list of source & target and using a loop to rsync them down" for the controller/load balancer is mainly for the sake of reducing code duplication and simplifying the code logic?

Reply:
Yes. Currently it has a lot of duplication. Please also double check other places in the code.